src/entities/bond.ts

```diff
@@ -90,6 +90,53 @@ export class Bond {
         return new Contract(this.address, BondAbi);
     }
 
+    trancheRedeemValue(
+        trancheAmount: CurrencyAmount<Token>,
+    ): CurrencyAmount<Token> {
+        const inputTranche: Tranche | undefined = this.tranches.find((t) =>
+            addressEquals(t.address, trancheAmount.currency.address),
+        );
+
+        invariant(inputTranche, 'Invalid input currency');
+
+        if (this.mature) {
+            return inputTranche.redeemValue(trancheAmount);
+        }
+
+        let remainingCollateral = this.totalCollateral;
+        for (let i = 0; i < this.tranches.length - 1; i++) {
+            const t = this.tranches[i];
+            const trancheCollateral = remainingCollateral.gt(t.totalSupply)
+                ? t.totalSupply
+                : remainingCollateral;
+            remainingCollateral = remainingCollateral.sub(trancheCollateral);
+
+            if (addressEquals(t.address, inputTranche.address)) {
+                return CurrencyAmount.fromRawAmount(
+                    this.collateral,
+                    trancheCollateral
+                        .mul(toBaseUnits(trancheAmount))
+                        .div(t.totalSupply)
+                        .toString(),
+                );
+            }
+        }
+
+        if (remainingCollateral.gt(0)) {
+            const zTranche = this.tranches[this.tranches.length - 1];
+            return CurrencyAmount.fromRawAmount(
+                this.collateral,
+                remainingCollateral
+                    .mul(toBaseUnits(trancheAmount))
+                    .div(zTranche.totalSupply)
+                    .toString(),
+            );
+        }
+
+        // this should never happen given the invariant above
```

Comment:
Currently this can happen when remainingCollateral = 0 for Z tranche and this query is made for the Z tranche balance

Reply:
oh.. true haha

Reply:
should return 0 in that case

Reply:
actually probably dont even need the conditional above... if remaining collateral is 0, we will already return 0